File: server-example/local/list.py
# auto install dependency
import subprocess, sys

for module, package in {'Crypto': 'pyCryptodome'}.items():
    try:
        __import__(module)
    except ModuleNotFoundError:
        if subprocess.call([sys.executable, '-m', 'pip', 'install', package]): raise

import json
import base64
import re
import hashlib
import binascii
from urllib.request import urlopen
from urllib.error import HTTPError
from datetime import datetime
from Crypto.Cipher import AES


def main():
    with open("cfg.json", 'r') as f:
        params = json.load(f)
        f.close()

    try:
        raw = urlopen(params["url"]).read().decode("utf-8")
        data = json.loads(raw)
    except HTTPError:
        print("Failed to read " + params["url"])
        return
    except json.decoder.JSONDecodeError:
        print("Failed to parse json " + raw)
        return

    if data.get("ok"):
        for row in data.get("messages", []):
            try:
                encoded = base64.urlsafe_b64decode(row["msg"])
            except binascii.Error as e:
                print(e, row["msg"])
                continue

            msg = decode(encoded, params["password"])
            if msg is None:
                print("Invalid password?")
                msg = encoded
            try:
                msg = str(msg, "utf-8")
            except UnicodeDecodeError:
                print("skip message " + row["msg"] + "\n")
                continue

            # 1ая строка - timestamp
            lines = re.split("\r\n", msg)
            if len(lines) > 0 and lines[0].isdigit():
                lines[0] = str(datetime.fromtimestamp(int(lines[0]) // 1000))
            print(*lines, "", sep="\n")


def decode(encoded, password):
    if password == "":
        return encoded
    try:
        # AES-128-CBC
        hashkey = hashlib.sha1(bytes(password, "utf-8")).digest()[0:16]
        iv = hashlib.sha1(hashkey).digest()[0:16]
        c = AES.new(hashkey, AES.MODE_CBC, iv)
        plaintext = c.decrypt(encoded)
        # remove padding
        return plaintext[:-plaintext[-1]]
    except Exception as e:
        return None


main()

File: server-example/local/list.php
<?php

function main(){
	
	$params = @json_decode(file_get_contents("cfg.json"), true);
	
	$data = @json_decode(file_get_contents($params['url']), true);
	
	if (!is_array($data))
		return;
	
	if (!@$data['ok'] || !is_array(@$data['messages']))
		return;
	
	foreach($data['messages'] as $row) {
		$encoded = url_safe_base64_decode($row['msg']);
		$msg = decode($encoded, $params['password']);
		if ($msg === false) {
			print("Invalid password?\r\n");
			$msg = $encoded;
		}

		$msg = @iconv("UTF-8", "cp866", $msg);
		if ($msg === false) {
			print("skip message " . $row['msg'] . "\r\n\r\n");
			continue;
		}
		
		$lines = preg_split("/\r\n/", $msg);
		if (count($lines) > 0 && is_numeric($lines[0])) {
			$dt = new DateTime("@" . intdiv($lines[0], 1000));
			$dt->setTimezone(new DateTimeZone('Europe/Moscow'));
			$lines[0] = $dt->format("Y-m-d H:i:s");
		}
		print(implode("\r\n", $lines) . "\r\n\r\n");
	}
};

function url_safe_base64_decode($data) {
	$data = base64_decode(str_replace(array('-','_'), array('+','/'), $data));
	return $data;	
}
 
function decode($encoded, $password) {
	if ($password == "")
		return $encoded;
	
	$cipher = "AES-128-CBC";

	$hashkey = substr(sha1($password, true), 0, 16);
	$iv = substr(sha1($hashkey, true), 0, 16);
	
	if ($plaintext = openssl_decrypt($encoded, $cipher, $hashkey, OPENSSL_RAW_DATA, $iv)) {
		return $plaintext;
	} else {
		return false;
	}
}


main();
